### PR TITLE
Add energy monitor

### PR DESCRIPTION
**Pull request by JiacongSun**
- Add energy monitor RTL and related submodule in a separate folder
- Add related testbenches
- Add the dtformat converter, the interface required by the analog island
- Add dtformat converter testbench

### CODE REVIEW

File: hw/rtl/dtformat_converter.sv [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+//
+// Parameters:
+// - N: number of data
+// - DATAW: bit precision of each data
+
+module dtformat_converter #(
+    parameter int N = 256,
+    parameter int DATAW = 4
+)(
+    // input logic clk_i, // input clock signal
```
Comment (clmcsn) [inline hw/rtl/dtformat_converter.sv:18]: Clean unused lines

File: hw/unit_tests/accumulator/tb_accumulator.sv [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        clear_i = 1;
+        #CLKCYCLE;
+        clear_i = 0;
+        clear_test_done = 1;
+    end
+
+    // Accumulation test
+    // use random data, random valid signal and random enable signal
+    initial begin
+        wait (clear_test_done == 1);
```
Comment (clmcsn) [inline hw/unit_tests/accumulator/tb_accumulator.sv:92]: A bit hacky! You might want to use `tasks `for future tb, they are the same thing but you can call in one single `initial `block

File: hw/unit_tests/adder_tree/tb_adder_tree.sv [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        .DATAW(DATAW)
+    ) adder_tree1 (
+        .data_i(a),
+        .sum_o(out)
+    );
+
+    // Testcases
+    // Test patterns for adder tree
+    logic signed [N*DATAW-1:0] test_a[NUM_TESTS] = '{
+        {N{8'sd0}}, // All zeros
```
Comment (clmcsn) [inline hw/unit_tests/adder_tree/tb_adder_tree.sv:36]: Not the most extensive test, but OK for now. We can work on a more practical and automated way to test blocks. E.g., use some python scripts to generate random input vectors

File: hw/unit_tests/adder_tree/tb_adder_tree.sv [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+`timescale 1ns / 1ps
+
+`ifndef DBG
+`define DBG 0
+`endif
+
+module tb_adder_tree;
+
+    localparam int N = 256; // number of inputs
```
Comment (clmcsn) [inline hw/unit_tests/adder_tree/tb_adder_tree.sv:13]: What if we have a different configuration?
Maybe add to the RTL blocks under which cases they have been tested (so we don't reuse them in other cases :))

File: hw/unit_tests/energy_monitor/tb_energy_monitor.sv [hunk 5]
```diff
@@ -0,0 +1,363 @@
+// Copyright 2025 KU Leuven.
```
Comment (clmcsn) [inline hw/unit_tests/energy_monitor/tb_energy_monitor.sv:1]: The testing procedure should be a bit more structured:
- Smoke test (simple input)
- Corner case (trigger special behavior)
- Random heavy testing (100 is not enough, you should go for 1 million!)

Each test should be contained in a separate file and easily inspectable by eye to verify its solidity. 
This should also be the case for the correctness check routines. I would finally write this in even another language (e.g., Python) to ensure even more that we are doing something that makes sense.
I am not asking to change it, but I would like to start a better way to test that will make our lives easier as well (tb are a pain)

File: hw/rtl/energy_monitor/partial_energy_calc.sv [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+module partial_energy_calc #(
+    parameter int BITJ = 4,
+    parameter int BITH = 4,
+    parameter int DATASPIN = 256,
+    parameter int SCALING_BIT = 5,
+    parameter int LOCAL_ENERGY_BIT = 16,
+    parameter int DATAJ = DATASPIN * BITJ
+    )(
+    // input logic clk_i, // input clock signal
```
[the remark this answers is not in the document]
Comment (clmcsn) [reply hw/rtl/energy_monitor/partial_energy_calc.sv:25]: Ok, I am not sure. The pipeline in your case is just at the input of the energy calculation module right? 
To be honest, I am not sure how well the retiming works and if it can move it from the input all the way to the computation.

To be sure, you would have to run synthesis on your block and see. What I am proposing is to modify your file _prehemptively_ (and avoid running synthesis) by adding pipeline stages e.g. between the "multiplication" and the adder tree. Any modification should be then extensively tested with different configurations (e.g., 0, 1 and 2 pipeline stages)

File: hw/rtl/energy_monitor/logic_ctrl.sv [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        SLEEP = 2'b00,
+        IDLE = 2'b01,
+        LOAD = 2'b10,
+        COMPUTE = 2'b11
+    } state_t;
+    state_t current_state, next_state;
+
+    assign config_ready_o = (current_state == IDLE);
+    assign spin_ready_o = (current_state == IDLE);
+    assign weight_ready_o = (current_state == LOAD);
```
[the remark this answers is not in the document]
Comment (JiacongSun) [reply hw/rtl/energy_monitor/logic_ctrl.sv:48]: In progress

File: hw/rtl/energy_monitor/partial_energy_calc.sv (hunk 6 above)
[the remark this answers is not in the document]
Comment (JiacongSun) [reply hw/rtl/energy_monitor/partial_energy_calc.sv:25]: I will not overdesign it and overtest it at this moment.

File: hw/rtl/energy_monitor/vector_caching.sv [hunk 8]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2025 KU Leuven.
```
[the remark this answers is not in the document]
Comment (JiacongSun) [reply hw/rtl/energy_monitor/vector_caching.sv:1]: The file has changed. Please check the new version.

File: hw/unit_tests/accumulator/tb_accumulator.sv (hunk 2 above)
Comment (JiacongSun) [reply hw/unit_tests/accumulator/tb_accumulator.sv:92]: I will keep it simple as this is a test for a leaf module.

File: hw/unit_tests/adder_tree/tb_adder_tree.sv (hunk 3 above)
Comment (JiacongSun) [reply hw/unit_tests/adder_tree/tb_adder_tree.sv:36]: I will keep it simple as this is a test for a leaf module.

File: hw/rtl/energy_monitor/partial_energy_calc.sv (hunk 6 above)
[the remark this answers is not in the document]
Comment (clmcsn) [reply hw/rtl/energy_monitor/partial_energy_calc.sv:25]: I don't think what I am asking is overdesign or overtesting :) 
You can see how the bachelor honours student did it last year:
https://github.com/KULeuven-MICAS/swirl/blob/qg-dev/hw/rtl/adder_tree.sv

If you don't want to make your design configurable, can you please run the synthesis to make sure we don't have problems later?

File: hw/unit_tests/energy_monitor/tb_energy_monitor.sv (hunk 5 above)
Comment (JiacongSun) [reply hw/unit_tests/energy_monitor/tb_energy_monitor.sv:1]: Good advice. To be done after finishing the entire RTL.

File: hw/rtl/energy_monitor/partial_energy_calc.sv (hunk 6 above)
[the remark this answers is not in the document]
Comment (JiacongSun) [reply hw/rtl/energy_monitor/partial_energy_calc.sv:25]: As you do not know the acceptable combinational depth under 500MHz, it will probably be okay without pipelines, as the multiplier is very simple. If it is not the case, it should be changed later on. But I do not have the bandwidth for now to verify this for a feature that may not be used later on. Finishing another module in time is the current priority.

I will ask for your help on the synthesis flow next week.

File: hw/unit_tests/adder_tree/tb_adder_tree.sv (hunk 4 above)
Comment (JiacongSun) [reply hw/unit_tests/adder_tree/tb_adder_tree.sv]: Files are removed.
